src/Hazelcast.Net/Sql/ISqlQueryResult.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Hazelcast.Sql
+{
+    /// <summary>
+    /// Represents SQL query (SELECT ...).
+    /// Provides methods to enumerate queried rows or cancel/dispose the query.
+    /// </summary>
+    public interface ISqlQueryResult: IAsyncEnumerator<SqlRow>, IAsyncDisposable
+    {
+        /// <summary>
+        /// <para>
+        /// Creates a one-off <see cref="IAsyncEnumerable{T}"/> around <see cref="SqlRow"/>s in row set.
+        /// </para>
+        /// <para>
+        /// Invoking this method again after enumeration has started, will throw <see cref="System.InvalidOperationException"/>.
+        /// </para>
+        /// <para>
+        /// Reusing obtained enumerable may lead to unpredictable results.
+        /// </para>
+        /// </summary>
+        /// <exception cref="ObjectDisposedException">SQL result was disposed.</exception>
+        /// <exception cref="InvalidOperationException">Enumeration has already started.</exception>
+        IAsyncEnumerable<SqlRow> EnumerateOnceAsync();
+
+        /// <summary>
+        /// <para>
+        /// Creates a one-off <see cref="IEnumerable{T}"/> around <see cref="SqlRow"/>s in row set.
+        /// </para>
+        /// <para>
+        /// Invoking this method again after enumeration has started, will throw <see cref="System.InvalidOperationException"/>.
+        /// </para>
+        /// <para>
+        /// Reusing obtained enumerable may lead to unpredictable results.
+        /// </para>
+        /// </summary>
+        /// <exception cref="ObjectDisposedException">SQL result was disposed.</exception>
+        /// <exception cref="InvalidOperationException">Enumeration has already started.</exception>
+        IEnumerable<SqlRow> EnumerateOnce();
```

Comment:
Do we need that sync method? Everything in our API is async, can't we simply have the async method?

Reply:
We already have this discussion ongoing, got hidden for some reason:
https://github.com/hazelcast/hazelcast-csharp-client/pull/475#discussion_r675435531

Reply:
I am having a hard time with these discussions getting hidden :( OK, 
* I don't want to keep the `GetAwaiter().GetResult()` thing in our code, that's an anti-pattern that can, and will, cause issues
* I would *love* to entirely remove everything about sync-enumerating
Let's slack-discuss it